### PR TITLE
Add IPS patching

### PR DESCRIPTION
**Pull request by zaksabeast**
This pull request:
- Patches the `.code` section of the exefs with an ips file located at `*.exefsdir/code.ips`
- Ensures the ips starts with "PATCH" before applying the patch
- Does not apply a patch if `*.exefsdir/code.bin` is present

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4353)

### CODE REVIEW

File: src/core/file_sys/ncch_container.cpp [hunk 1]
```diff
@@ -519,6 +534,25 @@ Loader::ResultStatus NCCHContainer::LoadOverrideExeFSSection(const char* name,
     return Loader::ResultStatus::ErrorNotUsed;
 }
 
+Loader::ResultStatus NCCHContainer::ApplyIPS(std::vector<u8>& ips, std::vector<u8>& buffer) {
+    u32 cursor = 5;
+    u32 patch_length = ips.size() - 3;
+    std::string ips_header(ips.begin(), ips.begin() + 5);
+
+    if (strcmp(ips_header.c_str(), "PATCH") != 0)
```
Comment (zhaowenlan1779) [inline src/core/file_sys/ncch_container.cpp]:
```suggestion
    if (ips_header != "PATCH")
```

File: src/core/file_sys/ncch_container.cpp [hunk 2]
```diff
@@ -519,6 +534,25 @@ Loader::ResultStatus NCCHContainer::LoadOverrideExeFSSection(const char* name,
     return Loader::ResultStatus::ErrorNotUsed;
 }
 
+Loader::ResultStatus NCCHContainer::ApplyIPS(std::vector<u8>& ips, std::vector<u8>& buffer) {
```
Comment (zhaowenlan1779) [inline src/core/file_sys/ncch_container.cpp]: This really does not need to return a value as you ignore it anyway. But please log if the ips is invalid, and make sure the data does not get messed up.

File: src/core/file_sys/ncch_container.cpp [hunk 3]
```diff
@@ -481,6 +481,21 @@ Loader::ResultStatus NCCHContainer::LoadSectionExeFS(const char* name, std::vect
...
+            std::string override_ips = filepath + ".exefsdir/code.ips";
+
+            if (FileUtil::Exists(override_ips) && strcmp(name, ".code") == 0) {
+                FileUtil::IOFile ips_file(override_ips, "rb");
+                std::size_t ips_file_size = ips_file.GetSize();
+                std::vector<u8> ips(ips_file_size);
+
+                if (ips_file.IsOpen() &&
+                    ips_file.ReadBytes(&ips[0], ips_file_size) == ips_file_size) {
+                    LOG_WARNING(Service_FS, "File {} patching code.bin", override_ips);
```
Comment (zhaowenlan1779) [inline src/core/file_sys/ncch_container.cpp]:
```suggestion
                    LOG_INFO(Service_FS, "File {} patching code.bin", override_ips);
```
likely.

File: src/core/file_sys/ncch_container.h [hunk 4]
```diff
@@ -249,6 +249,14 @@ class NCCHContainer {
      */
     Loader::ResultStatus ReadRomFS(std::shared_ptr<RomFSReader>& romfs_file);
 
+    /**
+     * Attempts to patch a buffer using an IPS
+     * @param ips Vector of the patches to apply
+     * @param buffer Vector to patch data into
+     * @return Void
```
Comment (wwylele) [inline src/core/file_sys/ncch_container.h]: nit: no need to have `@return` line for void return

**Comment by wwylele** [comment]
>While this is true, it has a pretty specific use case geared towards the AppLoader. However I can extract it if others agree it should be.

Make it static and private at least.

**Comment by zaksabeast** [comment]
I'm not sure what the proper way to resolve comments on reviewable.io is, but I believe all comments have been addressed.

**Comment by wwylele** [comment]
Will merge soon if no more comments

**Pull request merged**